dsgrid/spark/functions.py

```diff
@@ -0,0 +1,532 @@
+        """
+        return get_spark_session().sql(query)
+
+    return df1.join(df2, columns, how=how)
+
+
+def log_spark_conf(spark: SparkSession):
+    """Log the Spark configuration details."""
+    if not use_duckdb():
+        conf = spark.sparkContext.getConf().getAll()
+        conf.sort(key=lambda x: x[0])
+        logger.info("Spark conf: %s", "\n".join([f"{x} = {y}" for x, y in conf]))
+
+
+def prepare_timestamps_for_dataframe(timestamps: Iterable[datetime]) -> Iterable[datetime]:
+    """Apply necessary conversions of the timestamps for dataframe creation."""
+    if use_duckdb():
+        return [x.astimezone(ZoneInfo("UTC")) for x in timestamps]
+    return timestamps
+
+
+def read_csv(path: Path | str) -> DataFrame:
+    """Return a DataFrame from a CSV file, handling special cases with duckdb."""
+    spark = get_spark_session()
+    if use_duckdb():
+        path_ = path if isinstance(path, Path) else Path(path)
+        if path_.is_dir():
+            # path_str = str(path_) + "**/*.csv"
+            files = list(path_.glob("*.csv"))
+            assert len(files) == 1, files
+            path_str = str(files[0])
+        else:
+            path_str = str(path_)
+        # df = spark.read.csv(path_str, header=True)
+        # for field in df.schema:
+        #    if field.dataType is TimestampNTZType():
+        #        df = df.withColumn(field.name, F.col(field.name).cast(TimestampType()))
+        df = spark.createDataFrame(pd.read_csv(path_str))
+        if "timestamp" in df.columns:
+            # TODO duckdb: do something better
+            df = df.withColumn("timestamp", F.col("timestamp").cast(TimestampType()))
+        dup_cols = [x for x in df.columns if x.endswith(".1")]
+        if dup_cols:
+            msg = f"Detected a duplicate column in the dataset: {dup_cols}"
+            raise DSGInvalidDimension(msg)
+        return df
+    return spark.read.csv(str(path), header=True, inferSchema=True)
+
+
+def read_json(path: Path | str) -> DataFrame:
+    """Return a DataFrame from a JSON file, handling special cases with duckdb.
+
+    Warning: Use of this function with DuckDB is not efficient because it requires that we
+    convert line-delimited JSON to standard JSON.
+    """
+    spark = get_spark_session()
+    filename = str(path)
+    if use_duckdb():
+        with NamedTemporaryFile(suffix=".json") as f:
+            f.close()
+            data = load_line_delimited_json(path)
+            dump_data(data, f.name)
+            return spark.read.json(f.name)
+    return spark.read.json(filename, mode="FAILFAST")
+
+
+def read_parquet(path: Path | str) -> DataFrame:
+    path = Path(path) if isinstance(path, str) else path
+    spark = get_spark_session()
+    if path.is_file() or not use_duckdb():
+        df = spark.read.parquet(str(path))
+    else:
+        df = spark.read.parquet(f"{path}/**/*.parquet")
+    return df
+
+
+def select_expr(df: DataFrame, exprs: list[str]) -> DataFrame:
```

Comment:
I think these functions that toggle between pyspark and duckdB are great but the output are technically not the same, are they? Pyspark's can be daisy-chained, can the db's?

Reply:
They are consistent.